### PR TITLE
Deployment preparation: `TimelockAuthorizerTransitionMigrator`

### PR DESCRIPTION
**Pull request by jubeira**
# Deployment preparation: `TimelockAuthorizerTransitionMigrator`

Build-info: master @https://github.com/balancer-labs/balancer-v2-monorepo/commit/43e6405e189a8c69c24b396a88cef8388ea97e2e.
Supersedes #2151 (same, but rebased and squashed for convenience).

## Deployment preparation tasks checklist:

- [x] Deployments `README` is updated with the new task name and link
- [x] Task `README` has a description and links to the artifacts and expected output folders 
- [x] `build-info` is updated with artifacts compiled at the specified commit
- [x] Artifacts are generated from `build-info`
- [x] New task has a working fork test

## Code checklist:

- [x] The diff is legible and has no extraneous changes 
- [x] Complex code has been commented in tests and scripts
- [x] Fork test block number and test cases are appropriate for this deployment
- [x] The base branch is either `master`, or there's a description of how to merge

## Issue Resolution

See #1991

### CODE REVIEW

**Comment by TomAFrench** [comment]
> New task has a working fork test --> Adder V3 pending

This fork test will likely fail due to some of the permissions we're migrating having delays associated with granting them. We should update the migrator to handle these.

**Comment by jubeira** [comment]
> This fork test will likely fail due to some of the permissions we're migrating having delays associated with granting them. We should update the migrator to handle these.

Good point. The problem is not the adder itself (since the existing delays should apply only to the adder V2), but the gauge controller which is a singleton.

**Comment by jubeira** [comment]
Note: force-pushed to include #2184 while keeping this PR clean.

**Comment by nventuro** [comment]
I checked that the build info matches what is obtained by compiling `b270c7dbe44cc6fe07cf96f01166dc3ef3eb2e09`.

**Review by nventuro — APPROVED** [review]

File: pkg/deployments/tasks/scripts/20230130-ta-transition-migrator/input.ts [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+
+import { roles as mainnetRoles, delayedRoles as mainnetDelayedRoles } from './input/mainnet';
+import { RoleData } from './input/types';
+
+// Start: block that contains the transaction that deployed the `TimelockAuthorizer`.
+// https://etherscan.io/tx/0x20eb23f4393fd592240ec788f44fb9658cc6ef487b88398e9b76c910294c4eae
+// End: close to the current block at the time the `TimelockAuthorizerTransitionMigrator` is deployed.
+// It is expected that no roles were granted to the old authorizer after it.
+export const TRANSITION_START_BLOCK = 16085047;
+export const TRANSITION_END_BLOCK = 16477400;
```
Comment (TomAFrench) [inline pkg/deployments/tasks/scripts/20230130-ta-transition-migrator/input.ts:11]: Has this been updated?

File: pkg/deployments/tasks/scripts/20230130-ta-transition-migrator/test/task.fork.ts [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+    await migrator.migratePermissions();
+    for (const roleData of roles) {
+      expect(await newAuthorizer.hasPermission(roleData.role, roleData.grantee, roleData.target)).to.be.true;
+    }
+
+    for (const roleData of delayedRoles) {
+      expect(await newAuthorizer.hasPermission(roleData.role, roleData.grantee, roleData.target)).to.be.false;
+    }
+  });
+
```
Comment (TomAFrench) [inline pkg/deployments/tasks/scripts/20230130-ta-transition-migrator/test/task.fork.ts:71]: Can we chuck in a test or two here for what happens in the delay period? Should be mostly a copy paste from the normal tests.

**Review by TomAFrench — COMMENTED** [review]
Looks sensible. I can't verify the roles manually ATM but the code to check that it matches the onchain state looks fine.

Let's make sure the block numbers aren't stale and chuck in a couple of extra fork tests.

File: pkg/deployments/tasks/scripts/20230130-ta-transition-migrator/input.ts (hunk 1 above)
Comment (jubeira) [reply pkg/deployments/tasks/scripts/20230130-ta-transition-migrator/input.ts:11]: yes, it's a block from Jan-24.

**Pull request merged**